crates/bevy_utils/src/staging.rs

```diff
@@ -0,0 +1,1108 @@
+//!             Some(staged.secs_in_game)
+//!         } else {
+//!             self.cold.0.get(&id).map(|player| player.secs_in_game)
+//!         }?;
+//!         let additional = self
+//!             .staged
+//!             .additional_time_played
+//!             .get(&id)
+//!             .copied()
+//!             .unwrap_or_default();
+//!         Some(base + additional)
+//!     }
+//! }
+//!
+//! /// Allows mutable access to player data.
+//! trait PlayerAccessMut {
+//!     fn get_name_mut(&mut self, id: PlayerId) -> Option<impl DerefMut<Target = str>>;
+//!     fn add_secs_in_game(&mut self, id: PlayerId, secs: u32);
+//!     fn add(&mut self, name: String) -> PlayerId;
+//! }
+//!
+//! impl PlayerAccessMut for Players {
+//!     fn get_name_mut(&mut self, id: PlayerId) -> Option<impl DerefMut<Target = str>> {
+//!         self.0.get_mut(&id).map(|player| player.name.as_mut_str())
+//!     }
+//!
+//!     fn add_secs_in_game(&mut self, id: PlayerId, secs: u32) {
+//!         if let Some(player) = self.0.get_mut(&id) {
+//!             player.secs_in_game += secs;
+//!         }
+//!     }
+//!
+//!     fn add(&mut self, name: String) -> PlayerId {
+//!         let id = PlayerId(self.0.len() as u32);
+//!         self.0.insert(
+//!             id,
+//!             PlayerData {
+//!                 name,
+//!                 secs_in_game: 0,
+//!                 id,
+//!             },
+//!         );
+//!         id
+//!     }
+//! }
+//!
+//! impl PlayerAccessMut for Stager<'_, StagedPlayerChanges> {
+//!     fn get_name_mut(&mut self, id: PlayerId) -> Option<impl DerefMut<Target = str>> {
+//!         if !self.cold.0.contains_key(&id) && !self.staged.replacements.contains_key(&id) {
+//!             return None;
+//!         }
+//!
+//!         let player = self
+//!             .staged
+//!             .replacements
+//!             .entry(id)
+//!             .or_insert_with(|| self.cold.0.get(&id).cloned().unwrap());
+//!         Some(player.name.as_mut_str())
+//!     }
+//!
+//!     fn add_secs_in_game(&mut self, id: PlayerId, secs: u32) {
+//!         *self.staged.additional_time_played.entry(id).or_default() += secs;
+//!     }
+//!
+//!     fn add(&mut self, name: String) -> PlayerId {
+//!         let id = PlayerId((self.cold.0.len() + self.staged.replacements.len()) as u32);
+//!         self.staged.replacements.insert(
+//!             id,
+//!             PlayerData {
+//!                 name,
+//!                 secs_in_game: 0,
+//!                 id,
+//!             },
+//!         );
+//!         id
+//!     }
+//! }
+//!
+//! struct LockedNameStagedRef<'a> {
+//!     staged: RwLockReadGuard<'a, StagedPlayerChanges>,
+//!     // must be valid
+//!     id: PlayerId,
+//! }
+//!
+//! struct LockedNameColdRef<'a, T: StagableWritesCore<Staging = StagedPlayerChanges> + 'a> {
+//!     cold: T::ColdRef<'a>,
+//!     // must be valid
+//!     id: PlayerId,
+//! }
```

Comment:
Can we link both of these as workarounds to stabilization of https://doc.rust-lang.org/std/sync/struct.MappedRwLockReadGuard.html ? Just to make it clear that these are not strictly necessary.